update/common.go

```diff
@@ -0,0 +1,124 @@
+	//The below commented code should be removed, otherwise it would duplicate these miniblocks
+	//(especially the rewards miniblocks) which would be executed twice!
+	//for _, mbHdr := range epochStartMetaBlock.MiniBlockHeaders {
+	//	if mbHdr.SenderShardID != core.MetachainShardId || mbHdr.Type == block.PeerBlock {
+	//		continue
+	//	}
+	//	pendingMiniBlocks = append(pendingMiniBlocks, mbHdr)
+	//}
+
+	return pendingMiniBlocks, nil
+}
+
+// CreateNonceToHashMap creates a map of nonce to hash from all the given metaBlocks
+func CreateNonceToHashMap(unFinishedMetaBlocks map[string]*block.MetaBlock) map[uint64]string {
+	nonceToHashMap := make(map[uint64]string, len(unFinishedMetaBlocks))
+	for metaBlockHash, metaBlock := range unFinishedMetaBlocks {
+		nonceToHashMap[metaBlock.GetNonce()] = metaBlockHash
+	}
+
+	return nonceToHashMap
+}
+
+// ComputePendingMiniBlocksFromUnFinished computes all the pending miniBlocks from unFinished metaBlocks
+func ComputePendingMiniBlocksFromUnFinished(
+	epochStartShardData block.EpochStartShardData,
+	unFinishedMetaBlocks map[string]*block.MetaBlock,
+	nonceToHashMap map[uint64]string,
+	epochStartMetaBlockNonce uint64,
+) ([]block.MiniBlockHeader, error) {
+	pendingMiniBlocks := make([]block.MiniBlockHeader, 0)
+	pendingMiniBlocks = append(pendingMiniBlocks, epochStartShardData.PendingMiniBlockHeaders...)
+
+	firstPendingMetaBlock, ok := unFinishedMetaBlocks[string(epochStartShardData.FirstPendingMetaBlock)]
+	if !ok {
+		return nil, ErrWrongUnFinishedMetaHdrsMap
+	}
+
+	firstUnFinishedMetaBlockNonce := firstPendingMetaBlock.GetNonce()
+	for nonce := firstUnFinishedMetaBlockNonce + 1; nonce <= epochStartMetaBlockNonce; nonce++ {
+		metaBlockHash, exists := nonceToHashMap[nonce]
+		if !exists {
+			return nil, ErrWrongUnFinishedMetaHdrsMap
+		}
+
+		metaBlock, exists := unFinishedMetaBlocks[metaBlockHash]
+		if !exists {
+			return nil, ErrWrongUnFinishedMetaHdrsMap
+		}
+
+		pendingMiniBlocksFromMetaBlock := GetAllMiniBlocksWithDst(metaBlock, epochStartShardData.ShardID)
+		pendingMiniBlocks = append(pendingMiniBlocks, pendingMiniBlocksFromMetaBlock...)
+	}
+
+	return pendingMiniBlocks, nil
+}
+
+// GetAllMiniBlocksWithDst returns all miniBlock headers with the given destination from the given metaBlock
+func GetAllMiniBlocksWithDst(metaBlock *block.MetaBlock, destShardID uint32) []block.MiniBlockHeader {
+	mbHdrs := make([]block.MiniBlockHeader, 0)
+	for i := 0; i < len(metaBlock.ShardInfo); i++ {
+		if metaBlock.ShardInfo[i].ShardID == destShardID {
+			continue
+		}
+
+		for _, mbHdr := range metaBlock.ShardInfo[i].ShardMiniBlockHeaders {
+			if mbHdr.ReceiverShardID == destShardID && mbHdr.SenderShardID != destShardID {
+				mbHdrs = append(mbHdrs, mbHdr)
+			}
+		}
+	}
+
+	for _, mbHdr := range metaBlock.MiniBlockHeaders {
+		if mbHdr.ReceiverShardID == destShardID && mbHdr.SenderShardID != destShardID {
```

Comment:
```GO
isForMe := (mbHdr.ReceiverShardID == destShardID || mbHdr.ReceiverShardID == core.AllShardId) && mbHdr.SenderShardID != destShardID
````
as we have to take into account all shard ID value

Reply:
Actually core.AllShardId refers to PeerBlock right now, and PendingTransactionProcessor does not know to process this type of miniblock. More than this, in the commented code above from line 49 this type of miniblock was also avoided in the original code. Actually, the whole file is just a copy-paste of the original one with slightly renaming here and there.  We should discuss/analyze if this type of miniblock which could appear only in the start of epoch metablock (hardfork meta block) should be executed or not in the import phase or it would be executed in a normal way afterwards.

Reply:
Anyway I have deleted this method and I refactored the one from data/metablock.go -> GetOrderedCrossMiniblocksWithDst to be useful also here. This method takes in count also the PeerBlock miniblocks which should be skipped for now here, as PendingTransactionProcessor does not know to process them.

Reply:
I know, more mental mapping to have as to not use core.AllShardId on miniblocks other than peer miniblocks because there are some edgecases.